commcare_connect/opportunity/forms.py

```diff
@@ -1384,7 +1385,196 @@ def save(self, commit=True):
+        else:
+            self.fields["usd_currency"].widget.attrs.update(
+                {
+                    "x-ref": "currencyToggle",
+                    "x-on:change": "currency = $event.target.checked; convert(true)",
+                }
+            )
+
+        self.helper = FormHelper(self)
+
+        invoice_form_fields = self.invoice_form_fields()
+        if self.is_service_delivery:
+            invoice_form_fields.append(Field("notes"))
+
+        self.helper.layout = Layout(
+            *invoice_form_fields,
+            Div(
+                Submit("submit", "Submit", css_class="button button-md primary-dark"),
+                css_class="flex justify-end mt-4",
+            ),
+        )
+        self.helper.form_tag = False
+
+    def invoice_form_fields(self):
+        invoice_number_attrs = {}
+        invoice_number_attrs["readonly"] = "readonly"
+        first_row = [Field("invoice_number", **invoice_number_attrs)]
+
+        if self.is_service_delivery:
+            first_row.append(Field("title"))
+
+        second_row = [Field("date", **{"x-ref": "date", "x-on:change": "convert()"})]
+        if self.is_service_delivery:
+            second_row.append(Field("start_date"))
+            second_row.append(Field("end_date"))
+
+        third_row = [
+            Field(
+                "amount",
+                label=f"Amount ({self.opportunity.currency})",
+                **{
+                    "x-ref": "amount",
+                    "x-on:input.debounce.300ms": "convert()",
+                },
+            ),
+        ]
+        if self.is_service_delivery:
+            third_row.append(Field("amount_usd"))
+        else:
+            third_row.append(
+                Field(
+                    "usd_currency",
+                    css_class=CHECKBOX_CLASS,
+                    wrapper_class="flex p-4 justify-between rounded-lg bg-gray-100",
+                ),
+            )
+
+        return [
+            Div(
+                Div(*first_row, css_class="grid grid-cols-3 gap-6"),
+                Div(*second_row, css_class="grid grid-cols-3 gap-6"),
+                Div(
+                    *third_row,
+                    Div(css_id="converted-amount-wrapper", css_class="space-y-1 text-sm text-gray-500 mb-4"),
+                    css_class="grid grid-cols-3 gap-6",
+                ),
+                css_class="flex flex-col gap-4",
+            ),
+        ]
+
+    def generate_invoice_number(self):
+        return secrets.token_hex(5).upper()
+
+    def clean_invoice_number(self):
+        invoice_number = self.cleaned_data["invoice_number"]
+
+        if not invoice_number:
+            invoice_number = self.generate_invoice_number()
+
+        if PaymentInvoice.objects.filter(invoice_number=invoice_number).exists():
+            raise ValidationError(
+                "Please use a different invoice number",
```

Comment:
This will be a confusing message since they don't control the invoice number (we auto generate and don't let them edit). How do you feel about automatically regenerating a new one in this instance?